### ISSUE TITLE
Create service that polls the server for updates

### ISSUE THREAD

**Issue by thequbit** [hhroc/yellr-android#9]
This looks for new assignments, stories, notifications, and messages.  Once the service determines that the data is stale on the local store, then we will fire an intent that will push the new data to the app.

**Comment by thequbit** [comment]
We can use http://yellr.mycodespace.net/ for an end-point for the time being.  This is running HEAD of the https://github.com/hhroc/yellr-server code.

Wiki of API end-points: [https://github.com/hhroc/yellr/wiki/API-Documentation](https://github.com/hhroc/yellr/wiki/API-Documentation).

Note: that wiki page may be out of date, and I will need up update it.  It really needs to be moved over to yellr-server anyway.

**Comment by reustonium** [comment]
Might want to check out this description about how to notify from a service...
https://developer.android.com/guide/components/services.html#Notifications

I think you want an intent-service which is unbound...
https://developer.android.com/guide/components/services.html#LifecycleCallbacks

**Comment by thequbit** [comment]
Correct, definitely IntentService.

I've got this implemented and working for a get_assignments.json, and it returns it to the MainActivity thread when it's done.  Was NOT intuitive on how to do that, but I got it working.

So do you think I should I make a different IntentService (and corresponding Receiver class to do something with the response) for each end point?  We have ...

```
get_assignments.json
get_stories.json
get_notifications.json
get_messages.json
```

and then we'll need to perform an async for ...

```
upload_media.json
publish_post.json
```

I think the way I see it going is an IntentService for each of those, all with a corresponding Receiver.  I'm going to implement all of that unless you think otherwise @reustonium.  You have until midnight tonight to stop me, then I'm going to move forward with that plan, :).

**Comment by thequbit** [comment]
Commit with working JSON parsing:

https://github.com/thequbit/yellr-android/commit/3bdc249d0b2e651dbac473366555671fd01acc34

If you would like to take a look.

The WebWorkerIntentService would be renamed to GetAssignmentsIntentService in the above proposed model.

**Comment by thequbit** [comment]
This is closer to what I was suggesting:

https://github.com/thequbit/yellr-android/commit/77bc9e39a116fdfdeceac65c06c25368aa386c89

That is get_assignments.json and get-stories.json

Let me know whatcha think!

**Comment by reustonium** [comment]
I think you can probably get away with that design (a different service for each endpoint), but it might make more sense to have a single service for 'polling the backend' (which can launch 4 different intents).  And one service for uploading (media, and posts).

Looking at 
https://github.com/thequbit/yellr-android/commit/77bc9e39a116fdfdeceac65c06c25368aa386c89 that looks solid.

**Comment by thequbit** [comment]
Cool cool.  I think what I'll do is keep this architecture for the time being,  and keep functions separate enough that migration to a single service is less painful. 

Thanks @reustonium!!

**Comment by thequbit** [comment]
Implemented IntentService for get_assignments.json, get_stories.json, as well as publish_post.json (which calls upload_media.json from within it).

Need to implement get_notifications.json and get_messages.json.

Additionally we need to figure out how the responses from those calls then populate the fragments within the app.

I ... I think ... I think I'm enjoying writing java ... wat.

Edit:
Link to commit for implementation: https://github.com/thequbit/yellr-android/commit/02fecf7bd6dda831e7607413858eeffe3fbffee2
Additionally, I just put the calls for the IntentServices into the onCreate of the HomeActivity.  I suspect they will end up somewhere else eventually.

**Comment by thequbit** [comment]
Additionally, I pushed the APK to my phone (HTC One), and it worked as expected (published the post successfully to the server).

**Comment by reustonium** [comment]
that's awesome!  I plan on closing #2 and #14 today, will keep you posted.

**Comment by thequbit** [comment]
Awesomeness.  Drop a comment into each of the tickets, and I'll pull and
merge when you're ready.

On Wed, Feb 4, 2015 at 8:01 AM, Andy Ruestow notifications@github.com
wrote:

> that's awesome! I plan on closing #2
> https://github.com/hhroc/yellr-android/issues/2 and #14
> https://github.com/hhroc/yellr-android/issues/14 today, will keep you
> posted.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/hhroc/yellr-android/issues/9#issuecomment-72850380.

## 

_Tim Duffy_
http://timduffy.me/
(585)-210-8353
@arbiterofbits

**Comment by thequbit** [comment]
Here is what I think we need to move forward:

```
get_messages.json
get_notifications.json
get_assignments.json
get_stories.json
upload_media.json/publish_post.json
```

https://github.com/thequbit/yellr-android/commit/6914a07cd103a23ed3ff43e41278ad39d4d99a7d

These remain:

```
create_response_message.json
get_profile.json
```

We also can use 'server_info.json' to check to make sure we are using the correct version of the app with the correct version of the server.  I think we'll solve this issue with using nginx proxy_pass directives and keep old versions of APIs running until we don't see anyone using them anymore.

I hope to have those remaining two IntentServices done this evening.

Once complete, we need to figure out how and how often the four 'get' end-points (messages, notifications, assignments, and stories) are polled.  @reustonium I assume you have some kind of insight into this?  A timer in the Main Activity? Some kind of async messaging?  I'm a bit out of my wheelhouse again, ha.

**Comment by reustonium** [comment]
well the purpose of the service is to run in the background when the application is not open and to create a notification (which launches an intent with the story or assignment data bundled into it).

The App itself will make an aysnc call to the API directly (with retrofit?) each time that fragment is resumed (onResume()).

e.g. StoriesFragment will call the API in the `onResume()` method.  That code will send data to an `ArrayAdapter<Story>` which will populate the listview for each story (Title, text, images, etc).

This isn't a perfect example but check out how LunchPals does it...
https://github.com/reustonium/Lunch-Pals/blob/master/lunch-pals/src/main/java/com/reustonium/lunchpals/fragments/PangsListFragment.java

**Comment by thequbit** [comment]
Makes sense to me for the "I've opened the app, I need to get the the new things" case, however I don't understand how we'll handle the continued polling of the server.

Say, every 5 minutes I want to ping the server to see if we have new stories/assignments/messages/notifications.  How would we go about doing that?  Or am I missing something from your above example?

**Comment by thequbit** [comment]
IndentServices have been implemented to actually perform the HTTP method calls.  Now a service needs to be implemented that will periodically call those IndentServices and repopulate the UI as needed.

**Comment by thequbit** [comment]
I thought I closed with with #87 , however @reustonium showed that is not the case.  Will work to implement AlarmManager implementation this evening, and resubmit a PR.

**Comment by thequbit** [comment]
This is implemented, and can be closed.  There is #143 that seems to exist on some phones??  Closing for now.
